components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/XACMLScopeValidator.java

```diff
@@ -0,0 +1,210 @@
+import java.util.ArrayList;
+import java.util.List;
+import javax.xml.stream.XMLStreamException;
+
+
+/**
+ * The Scope Validation implementation. This uses XACML policies to evaluate scope validation defined by the user.
+ */
+public class XACMLScopeValidator extends OAuth2ScopeValidator {
+
+    private static final String DECISION_XPATH = "//ns:Result/ns:Decision/text()";
+    private static final String XACML_NS_PREFIX = "ns";
+    private static final String ACTION_VALIDATE = "token_validation";
+    private static final String RULE_EFFECT_PERMIT = "Permit";
+    private static final String RULE_EFFECT_NOT_APPLICABLE = "NotApplicable";
+
+    private static final String ACTION_CATEGORY = "http://wso2.org/identity/identity-action";
+    private static final String SP_CATEGORY = "http://wso2.org/identity/sp";
+    private static final String USER_CATEGORY = "http://wso2.org/identity/user";
+    private static final String SCOPE_CATEGORY = "http://wso2.org/identity/oauth-scope";
+
+    private static final String AUTH_ACTION_ID = ACTION_CATEGORY + "/action-name";
+    private static final String SP_NAME_ID = SP_CATEGORY + "/sp-name";
+    private static final String USERNAME_ID = USER_CATEGORY + "/username";
+    private static final String USER_STORE_ID = USER_CATEGORY + "/user-store-domain";
+    private static final String USER_TENANT_DOMAIN_ID = USER_CATEGORY + "/user-tenant-domain";
+    private static final String SCOPE_ID = SCOPE_CATEGORY + "/scope-name";
+
+    private Log log = LogFactory.getLog(XACMLScopeValidator.class);
+
+    @Override
+    public boolean validateScope(AccessTokenDO accessTokenDO, String resource) throws IdentityOAuth2Exception {
+
+        if (accessTokenDO.getAuthzUser() == null) {
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("There is no authorized user for access token id %s ",
+                        accessTokenDO.getTokenId()));
+            }
+            return false;
+        }
+
+        boolean isValidated = false;
+        try {
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("In XACML based scope validation flow for access token of consumer key : %s ",
+                        accessTokenDO.getConsumerKey()));
+            }
+            String consumerKey = accessTokenDO.getConsumerKey();
+            OAuthAppDO authApp = OAuth2Util.getAppInformationByClientId(consumerKey);
+
+            RequestDTO requestDTO = createRequestDTO(accessTokenDO, authApp, resource);
+            RequestElementDTO requestElementDTO = PolicyCreatorUtil.createRequestElementDTO(requestDTO);
+            String requestString = PolicyBuilder.getInstance().buildRequest(requestElementDTO);
+
+            if (log.isDebugEnabled()) {
+                log.debug("XACML scope validation request :\n" + requestString);
+            }
+            FrameworkUtils.startTenantFlow(accessTokenDO.getAuthzUser().getTenantDomain());
```

Comment:
If an exception is thrown in a line above this, let's say in line 97 tenant flow won't be started. However, the finally block will be executed and attempt to end a tenant flow would cause an error.